// @wundergraph/sdk/dist/dotgraphqlconfig/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDotGraphQLConfig = void 0;
const path_1 = __importDefault(require("path"));
const generateDotGraphQLConfig = (wunderGraphConfig, options) => {
    let config = {
        projects: {}
    };
    const app = wunderGraphConfig.application;
    const deployment = "main";
    config.projects[app.name] = {
        name: app.name,
        schemaPath: path_1.default.join(options.nested ? ".wundergraph" : "", "generated", `wundergraph.${app.name}.schema.graphql`),
        extensions: {
            endpoints: {
                [app.name]: {
                    introspect: false,
                    url: `${options.baseURL}/${app.name}/${deployment}/graphql`,
                    headers: {
                        "user-agent": "WunderGraph Client"
                    }
                },
            }
        }
    };
    return config;
};
exports.generateDotGraphQLConfig = generateDotGraphQLConfig;
