// @wundergraph/sdk/dist/definition/merge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeBaseSchema = exports.baseSchema = exports.mergeApis = void 0;
const index_1 = require("./index");
const graphql_1 = require("graphql");
const schema_1 = require("@graphql-tools/schema");
const wundernode_config_1 = require("@wundergraph/protobuf/dist/wundernode_config");
const mergeApis = (roles, ...apis) => {
    const dataSources = apis
        .map(api => api.DataSources)
        .reduce((previousValue, currentValue) => ([...previousValue, ...currentValue]));
    const fields = mergeApiFields(apis);
    const types = mergeTypeConfigurations(apis);
    const schema = mergeApiSchemas(roles, apis, dataSources, fields);
    return new index_1.Api(schema, dataSources, fields, types);
};
exports.mergeApis = mergeApis;
const mergeApiFields = (apis) => {
    const fields = [];
    apis.map(a => a.Fields).reduce((previousValue, currentValue) => ([...previousValue, ...currentValue])).forEach(f => {
        const existing = fields.find(existing => existing.typeName === f.typeName && existing.fieldName === f.fieldName);
        if (!existing) {
            fields.push(f);
            return;
        }
        if (f.requiresFields) {
            existing.requiresFields = [...new Set([...existing.requiresFields || [], ...f.requiresFields])];
        }
        if (f.argumentsConfiguration) {
            existing.argumentsConfiguration = existing.argumentsConfiguration || [];
            f.argumentsConfiguration.forEach(a => {
                const exists = existing.argumentsConfiguration.find(e => e.name === a.name) !== undefined;
                if (exists) {
                    return;
                }
                existing.argumentsConfiguration.push(a);
            });
        }
    });
    return fields;
};
const mergeTypeConfigurations = (apis) => {
    const out = [];
    apis.forEach(api => {
        api.Types.forEach(conf => {
            const exists = out.find(existing => existing.typeName === conf.typeName) !== undefined;
            if (exists) {
                return;
            }
            out.push(conf);
        });
    });
    return out;
};
exports.baseSchema = `
directive @fromClaim(
  name: Claim
) on VARIABLE_DEFINITION

enum Claim {
  EMAIL
  EMAIL_VERIFIED
  NAME
  NICKNAME
  LOCATION
  PROVIDER
}

directive @hooksVariable on VARIABLE_DEFINITION

directive @jsonSchema (
  """
  The value of both of these keywords MUST be a string.

  Both of these keywords can be used to decorate a user interface with
  information about the data produced by this user interface.  A title
  will preferably be short, whereas a description will provide
  explanation about the purpose of the instance described by this
  schema.
  """
  title: String
  """
  The value of both of these keywords MUST be a string.

  Both of these keywords can be used to decorate a user interface with
  information about the data produced by this user interface.  A title
  will preferably be short, whereas a description will provide
  explanation about the purpose of the instance described by this
  schema.
  """
  description: String
  """
  The value of "multipleOf" MUST be a number, strictly greater than 0.

  A numeric instance is valid only if division by this keyword's value
  results in an integer.
  """
  multipleOf: Int
  """
  The value of "maximum" MUST be a number, representing an inclusive
  upper limit for a numeric instance.

  If the instance is a number, then this keyword validates only if the
  instance is less than or exactly equal to "maximum".
  """
  maximum: Int
  """
  The value of "exclusiveMaximum" MUST be number, representing an
  exclusive upper limit for a numeric instance.

  If the instance is a number, then the instance is valid only if it
  has a value strictly less than (not equal to) "exclusiveMaximum".
  """
  exclusiveMaximum: Int
  """
  The value of "minimum" MUST be a number, representing an inclusive
  lower limit for a numeric instance.

  If the instance is a number, then this keyword validates only if the
  instance is greater than or exactly equal to "minimum".
  """
  minimum: Int
  """
  The value of "exclusiveMinimum" MUST be number, representing an
  exclusive lower limit for a numeric instance.

  If the instance is a number, then the instance is valid only if it
  has a value strictly greater than (not equal to) "exclusiveMinimum".
  """
  exclusiveMinimum: Int
  """
  The value of this keyword MUST be a non-negative integer.

  A string instance is valid against this keyword if its length is less
  than, or equal to, the value of this keyword.

  The length of a string instance is defined as the number of its
  characters as defined by RFC 7159 [RFC7159].
  """
  maxLength: Int
  """
  The value of this keyword MUST be a non-negative integer.

  A string instance is valid against this keyword if its length is
  greater than, or equal to, the value of this keyword.

  The length of a string instance is defined as the number of its
  characters as defined by RFC 7159 [RFC7159].

  Omitting this keyword has the same behavior as a value of 0.
  """
  minLength: Int
  """
  The value of this keyword MUST be a string.  This string SHOULD be a
  valid regular expression, according to the ECMA 262 regular
  expression dialect.

  A string instance is considered valid if the regular expression
  matches the instance successfully.  Recall: regular expressions are
  not implicitly anchored.
  """
  pattern: String
  """
  The value of this keyword MUST be a non-negative integer.

  An array instance is valid against "maxItems" if its size is less
  than, or equal to, the value of this keyword.
  """
  maxItems: Int
  """
  The value of this keyword MUST be a non-negative integer.

  An array instance is valid against "minItems" if its size is greater
  than, or equal to, the value of this keyword.

  Omitting this keyword has the same behavior as a value of 0.
  """
  minItems: Int
  """
  The value of this keyword MUST be a boolean.

  If this keyword has boolean value false, the instance validates
  successfully.  If it has boolean value true, the instance validates
  successfully if all of its elements are unique.

  Omitting this keyword has the same behavior as a value of false.
  """
  uniqueItems: Boolean
  commonPattern: COMMON_REGEX_PATTERN
) on VARIABLE_DEFINITION

enum COMMON_REGEX_PATTERN {
    EMAIL
    DOMAIN
}
`;
const roleSchema = (roles) => `
directive @rbac(
  "the user must match all roles"
  requireMatchAll: [WG_ROLE]
  "the user must match at least one of the roles"
  requireMatchAny: [WG_ROLE]
  "the user must not match all of the roles"
  denyMatchAll: [WG_ROLE]
  "the user must not match any of the roles"
  denyMatchAny: [WG_ROLE]
) on QUERY | MUTATION | SUBSCRIPTION

enum WG_ROLE {
    ${roles.join(" ")}
}
`;
const uuidSchema = `
"""
The directive @injectGeneratedUUID injects a generated UUID into the variable.
This variable MUST be a string.
At the same time, it removes the variable from the input definition,
disallowing the user to supply it.

This means, the UUID is 100% generated server-side and can be considered untempered.
"""
directive @injectGeneratedUUID on VARIABLE_DEFINITION
`;
const injectEnvironmentVariableSchema = `
"""
The directive @injectEnvironmentVariable allows you to inject an environment variable into the variable definition.
"""
directive @injectEnvironmentVariable (
    name: String!
) on VARIABLE_DEFINITION
`;
const dateTimeSchema = `
"""
The directive @injectCurrentDateTime injects a DateTime string of the current date and time into the variable.
This variable MUST be a string compatible scalar. 

The default format, is: ISO 8601
If no format is chosen, the default format is used.
Custom formats are allowed by specifying a format conforming to the Golang specification for specifying a date time format.
"""
directive @injectCurrentDateTime (
    format: WunderGraphDateTimeFormat = ISO8601
    """customFormat must conform to the Golang specification for specifying a date time format"""
    customFormat: String
) on VARIABLE_DEFINITION

enum WunderGraphDateTimeFormat {
    "2006-01-02T15:04:05-0700"
    ISO8601
    "Mon Jan _2 15:04:05 2006"
    ANSIC
    "Mon Jan _2 15:04:05 MST 2006"
    UnixDate
    "Mon Jan 02 15:04:05 -0700 2006"
    RubyDate
    "02 Jan 06 15:04 MST"
    RFC822
    "02 Jan 06 15:04 -0700"
    RFC822Z
    "Monday, 02-Jan-06 15:04:05 MST"
    RFC850
    "Mon, 02 Jan 2006 15:04:05 MST"
    RFC1123
    "Mon, 02 Jan 2006 15:04:05 -0700"
    RFC1123Z
    "2006-01-02T15:04:05Z07:00"
    RFC3339
    "2006-01-02T15:04:05.999999999Z07:00"
    RFC3339Nano
    "3:04PM"
    Kitchen
    "Jan _2 15:04:05"
    Stamp
    "Jan _2 15:04:05.000"
    StampMilli
    "Jan _2 15:04:05.000000"
    StampMicro
    "Jan _2 15:04:05.000000000"
    StampNano
}
`;
const internalSchema = `
"""
The @internalOperation Directive marks an Operation as internal.
By doing so, the Operation is no longer accessible from the public API.
It can only be accessed by internal services, like hooks.
"""
directive @internalOperation on QUERY | MUTATION | SUBSCRIPTION 
`;
const exportSchema = `

"""
The @export directive instructs the Execution Planner to export the field during the execution into the variable of the 'as' argument.
As the execution is depth first, a field can only be used after it has been exported.
Additionally, a field can only be used after using the '_join' field or on a different data source.
It's not possible to export a field and use it in for the same data source.

Note that the @export directive only works on fields that return a single value.
It's not possible to export a list or object field.
"""
directive @export (
    """
    The argument 'as' is the name of the variable to export the field to.
    """
    as: String!
) on FIELD

"""
The directive @internal marks a variable definition as internal so that clients can't access it.
The field is also not visible in the public API.
It's only being used as an internal variable to export fields into.
"""
directive @internal on VARIABLE_DEFINITION
`;
const mergeApiSchemas = (roles, apis, dataSources, fields) => {
    const graphQLSchemas = apis.map((api, i) => {
        return (0, graphql_1.buildSchema)(api.Schema, {
            assumeValidSDL: true,
        });
    });
    graphQLSchemas.push((0, graphql_1.buildSchema)(exports.baseSchema, {
        assumeValidSDL: true
    }));
    if (roles.length) {
        graphQLSchemas.push((0, graphql_1.buildSchema)(roleSchema(roles), { assumeValidSDL: true }));
    }
    graphQLSchemas.push((0, graphql_1.buildSchema)(dateTimeSchema, { assumeValidSDL: true }));
    graphQLSchemas.push((0, graphql_1.buildSchema)(uuidSchema, { assumeValidSDL: true }));
    graphQLSchemas.push((0, graphql_1.buildSchema)(internalSchema, { assumeValidSDL: true }));
    graphQLSchemas.push((0, graphql_1.buildSchema)(injectEnvironmentVariableSchema, { assumeValidSDL: true }));
    graphQLSchemas.push((0, graphql_1.buildSchema)(exportSchema, { assumeValidSDL: true }));
    const mergedGraphQLSchema = (0, schema_1.mergeSchemas)({
        schemas: graphQLSchemas,
        assumeValid: true,
    });
    const printed = (0, graphql_1.printSchema)(mergedGraphQLSchema);
    const ast = (0, graphql_1.parse)(printed);
    const filtered = (0, graphql_1.visit)(ast, {
        ObjectTypeDefinition: node => {
            var _a;
            if (node.name.value.startsWith("__")) {
                return null;
            }
            switch (node.name.value) {
                case "Entity":
                    return null;
            }
            const hasJoinField = ((_a = node.fields) === null || _a === void 0 ? void 0 : _a.find(field => field.name.value === "_join")) !== undefined;
            const isRootType = node.name.value === "Query" || node.name.value === "Mutation" || node.name.value === "Subscription";
            if (hasJoinField || isRootType) {
                return;
            }
            const custom = {
                data: {
                    kind: wundernode_config_1.ConfigurationVariableKind.STATIC_CONFIGURATION_VARIABLE,
                    staticVariableContent: "{}",
                    placeholderVariableName: "",
                    environmentVariableDefaultValue: "",
                    environmentVariableName: "",
                },
            };
            dataSources.push({
                Kind: wundernode_config_1.DataSourceKind.STATIC,
                RootNodes: [
                    {
                        typeName: node.name.value,
                        fieldNames: ["_join"]
                    }
                ],
                Custom: custom,
                ChildNodes: [],
                Directives: [],
            });
            fields.push({
                typeName: node.name.value,
                fieldName: "_join",
                disableDefaultFieldMapping: true,
                path: ["_join"],
                requiresFields: [],
                argumentsConfiguration: [],
            });
            const updated = {
                ...node,
                fields: [
                    ...node.fields || [],
                    {
                        kind: "FieldDefinition",
                        type: {
                            kind: "NonNullType",
                            type: {
                                kind: "NamedType",
                                name: {
                                    kind: "Name",
                                    value: "Query",
                                }
                            }
                        },
                        name: {
                            kind: "Name",
                            value: "_join",
                        },
                    },
                ]
            };
            return updated;
        },
        UnionTypeDefinition: node => {
            if (node.name.value.startsWith("__")) {
                return null;
            }
        },
        ScalarTypeDefinition: node => {
            if (node.name.value.startsWith("__")) {
                return null;
            }
        },
        FieldDefinition: node => {
            if (node.name.value.startsWith("__")) {
                return null;
            }
        },
        DirectiveDefinition: node => {
            switch (node.name.value) {
                case "key":
                case "extends":
                case "external":
                case "requires":
                case "provides":
                    return null;
            }
        }
    });
    const withoutEmptyDescription = removeEmptyDescriptions(filtered);
    return (0, graphql_1.print)(withoutEmptyDescription);
};
const removeEmptyDescriptions = (astNode) => {
    return (0, graphql_1.visit)(astNode, {
        enter: node => {
            switch (node.kind) {
                case "ObjectTypeDefinition":
                case "ScalarTypeDefinition":
                case "InterfaceTypeDefinition":
                case "UnionTypeDefinition":
                case "FieldDefinition":
                    if (node.description && node.description.value === "") {
                        return {
                            ...node,
                            description: undefined,
                        };
                    }
            }
        }
    });
};
const wunderGraphDirectives = [
    "internalOperation",
    "injectEnvironmentVariable",
    "fromClaim",
    "hooksVariable",
    "jsonSchema",
    "rbac",
    "injectGeneratedUUID",
    "injectCurrentDateTime",
    "internalOperation",
    "export",
];
const wunderGraphEnums = [
    "Claim",
    "COMMON_REGEX_PATTERN",
    "WG_ROLE",
    "WunderGraphDateTimeFormat"
];
const removeBaseSchema = (schema) => {
    const document = (0, graphql_1.parse)(schema);
    const filtered = (0, graphql_1.visit)(document, {
        ObjectTypeDefinition: node => {
            var _a;
            const fields = (_a = node.fields) === null || _a === void 0 ? void 0 : _a.filter(field => field.name.value !== "_join");
            return {
                ...node,
                fields,
            };
        },
        DirectiveDefinition: node => {
            if (wunderGraphDirectives.includes(node.name.value)) {
                return null;
            }
        },
        EnumTypeDefinition: node => {
            if (wunderGraphEnums.includes(node.name.value)) {
                return null;
            }
        }
    });
    return (0, graphql_1.print)(filtered);
};
exports.removeBaseSchema = removeBaseSchema;
