// @wundergraph/sdk/dist/codegen/templates/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = require("./typescript");
const client_1 = require("./typescript/client");
const react_1 = require("./typescript/react");
const operations_1 = require("./typescript/operations");
const linkbuilder_1 = require("./typescript/linkbuilder");
const jsonschema_1 = require("./typescript/jsonschema");
const forms_1 = require("./typescript/forms");
const wundergraphHooks_1 = require("./typescript/wundergraphHooks");
const authentication_1 = require("./markdown/authentication");
const typescriptModels = [
    new typescript_1.TypeScriptInputModels(),
    new typescript_1.TypeScriptInternalInputModels(),
    new typescript_1.TypeScriptInjectedInputModels(),
    new typescript_1.TypeScriptResponseModels(),
    new typescript_1.TypeScriptResponseDataModels(),
    new authentication_1.AuthenticationProviderConfiguration(),
];
const typescriptAll = [
    ...typescriptModels,
    new wundergraphHooks_1.WunderGraphTypeScriptHooks(),
    new operations_1.Operations(),
    new jsonschema_1.JsonSchema(),
    new forms_1.Forms(),
];
const templates = {
    typescript: {
        // models generates all models for input and output
        models: typescriptModels,
        inputModels: new typescript_1.TypeScriptInputModels(),
        responseModels: new typescript_1.TypeScriptResponseModels(),
        client: new client_1.TypeScriptClient(),
        all: [...typescriptAll, new client_1.TypeScriptClient()],
        wunderGraphHooks: new wundergraphHooks_1.WunderGraphTypeScriptHooks(),
        react: [
            ...typescriptAll,
            new client_1.TypeScriptClient(),
            new react_1.TypescriptReactProvider(),
            new react_1.TypescriptReactHooks(),
        ],
        reactNative: [
            ...typescriptAll,
            new client_1.TypeScriptClient(true),
            new react_1.TypescriptReactNativeProvider(),
            new react_1.TypescriptReactHooks(true),
        ],
        operations: new operations_1.Operations(),
        linkBuilder: new linkbuilder_1.TypeScriptLinkBuilder(),
        jsonSchema: new jsonschema_1.JsonSchema(),
        forms: new forms_1.Forms(),
    }
};
exports.default = templates;
