// @wundergraph/sdk/dist/codegen/templates/typescript/jsonschema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonSchema = void 0;
const index_1 = require("./index");
const handlebars_1 = __importDefault(require("handlebars"));
// @ts-ignore
const jsonschema_ts_handlebars_1 = __importDefault(require("./jsonschema.ts.handlebars"));
class JsonSchema {
    generate(config) {
        const model = {
            operations: config.application.Operations.map(op => ({
                name: op.Name,
                inputSchema: JSON.stringify(op.VariablesSchema),
                outputSchema: JSON.stringify(op.ResponseSchema),
            }))
        };
        const input = (0, index_1.loadFile)(jsonschema_ts_handlebars_1.default);
        const tmpl = handlebars_1.default.compile(input);
        const content = tmpl(model);
        return Promise.resolve([
            {
                path: "jsonschema.ts",
                content: (0, index_1.formatTypeScript)(content),
                doNotEditHeader: true,
            }
        ]);
    }
}
exports.JsonSchema = JsonSchema;
