// @wundergraph/sdk/dist/codegen/templates/typescript/wundergraphHooks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WunderGraphTypeScriptHooks = void 0;
const handlebars_1 = __importDefault(require("handlebars"));
const index_1 = require("./index");
const wundernode_config_1 = require("@wundergraph/protobuf/dist/wundernode_config");
const client_1 = require("./client");
// @ts-ignore
const wundergraphHooks_ts_handlebars_1 = __importDefault(require("./wundergraphHooks.ts.handlebars"));
class WunderGraphTypeScriptHooks {
    generate(config) {
        const input = (0, index_1.loadFile)(wundergraphHooks_ts_handlebars_1.default);
        const tmpl = handlebars_1.default.compile(input);
        const _queries = (0, client_1.operations)(config.application, wundernode_config_1.OperationType.QUERY, false);
        const _internalQueries = (0, client_1.operations)(config.application, wundernode_config_1.OperationType.QUERY, true);
        const _mutations = (0, client_1.operations)(config.application, wundernode_config_1.OperationType.MUTATION, false);
        const _internalMutations = (0, client_1.operations)(config.application, wundernode_config_1.OperationType.MUTATION, true);
        const content = tmpl({
            applicationName: config.application.Name,
            modelImports: (0, client_1.modelImports)(config.application, true),
            queries: _queries,
            hasQueries: _queries.length !== 0,
            mutations: _mutations,
            hasMutations: _mutations.length !== 0,
            internalQueries: _internalQueries,
            hasInternalQueries: _internalQueries.length !== 0,
            internalMutations: _internalMutations,
            hasInternalMutations: _internalMutations.length !== 0,
            serverPort: '9992',
            roleDefinitions: config.authentication.roles.map(role => "\"" + role + "\"").join(" | "),
        });
        return Promise.resolve([
            {
                path: "wundergraph.hooks.configuration.ts",
                content: (0, index_1.formatTypeScript)(content),
                doNotEditHeader: true,
            }
        ]);
    }
    dependencies() {
        return [
            new index_1.TypeScriptInputModels(),
            new index_1.TypeScriptResponseModels(),
        ];
    }
}
exports.WunderGraphTypeScriptHooks = WunderGraphTypeScriptHooks;
