// @wundergraph/sdk/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.authProviders = exports.cors = exports.EnvironmentVariable = exports.resolveIntegration = exports.configureWunderGraphApplication = exports.Api = exports.createMockApi = exports.introspect = exports.Application = exports.templates = void 0;
var templates_1 = require("./codegen/templates");
Object.defineProperty(exports, "templates", { enumerable: true, get: function () { return __importDefault(templates_1).default; } });
var definition_1 = require("./definition");
Object.defineProperty(exports, "Application", { enumerable: true, get: function () { return definition_1.Application; } });
Object.defineProperty(exports, "introspect", { enumerable: true, get: function () { return definition_1.introspect; } });
Object.defineProperty(exports, "createMockApi", { enumerable: true, get: function () { return definition_1.createMockApi; } });
Object.defineProperty(exports, "Api", { enumerable: true, get: function () { return definition_1.Api; } });
var configure_1 = require("./configure");
Object.defineProperty(exports, "configureWunderGraphApplication", { enumerable: true, get: function () { return configure_1.configureWunderGraphApplication; } });
Object.defineProperty(exports, "resolveIntegration", { enumerable: true, get: function () { return configure_1.resolveIntegration; } });
Object.defineProperty(exports, "EnvironmentVariable", { enumerable: true, get: function () { return configure_1.EnvironmentVariable; } });
var cors_1 = require("./cors");
Object.defineProperty(exports, "cors", { enumerable: true, get: function () { return __importDefault(cors_1).default; } });
var authentication_1 = require("./configure/authentication");
Object.defineProperty(exports, "authProviders", { enumerable: true, get: function () { return authentication_1.authProviders; } });
